### PR TITLE
qa/tasks: improvements in vstart_runner.py and mount.py

### PR DESCRIPTION
**Pull request by rishabh-d-dave**
Brief summary for this PR:
* makes crashing optional (but default) for `run_shell()`
* adds methods that can run a command with a user other than
  the current user (including as the root user)
* allows running commands with `sudo` (right now `run_shell` elides `sudo`s
  from command arguments)
* makes `run_shell()` accept command arguments as string (right now it only
  accepts a list of args)
* allows passing stdin to `run_shell()`
* adds a method to create an empty file as any user
* adds wrappers for `vstart_runner.py`'s `run()` that is tuned to *test*
  a command rather than executing a command with the hope for success.

These modifications were originally made while I was writing PR #26477.

### CODE REVIEW

**Comment by rishabh-d-dave** [comment]
I don't know if there are any docs around around all the `vstart_runner.py` and `mount.py`. If we do have them, I'll be happy to update them as well.

**Comment by batrick** [comment]
> I don't know if there are any docs around around all the vstart_runner.py and mount.py. If we do have them, I'll be happy to update them as well.

There are no docs yet for `vstart_runner.py` but if you created a doc you would earn major kudos from me. The only documentation that exists is what's found on the mailing list, including my recent post here: https://lists.ceph.io/hyperkitty/list/ceph-fs@ceph.io/thread/GXUVNFMYXVIGWHG3N2NDFJTUCQTUS26G/

The right place to add a doc would be doc/dev/cephfs

File: qa/tasks/vstart_runner.py [hunk 1]
```diff
@@ -236,10 +236,19 @@ def put_file(self, src, dst, sudo=False):
 
     def run(self, args, check_status=True, wait=True,
             stdout=None, stderr=None, cwd=None, stdin=None,
-            logger=None, label=None, env=None, timeout=None):
+            logger=None, label=None, env=None, timeout=None, omit_sudo=True):
+        if str(type(args)).find('str') != -1:
```
Comment (batrick) [inline qa/tasks/vstart_runner.py]: why not `isinstance(args, str)`?

File: qa/tasks/vstart_runner.py [hunk 2]
```diff
@@ -418,13 +427,92 @@ def get_keyring_path(self):
...
+
+    def testcmd(self, args, wait=True, stdin=None, check_status=False, omit_sudo=True):
+        # FIXME maybe should add a pwd arg to teuthology.orchestra so that
+        # the "cd foo && bar" shenanigans isn't needed to begin with and
+        # then we wouldn't have to special case this
+        return self.client_remote.run(args, wait=wait, cwd=self.mountpoint,
+                                      stdin=stdin, check_status=check_status,
+                                      omit_sudo=omit_sudo)
+
+    def testcmd_as_user(self, args, user, pwd=None, wait=True, check_status=False):
```
Comment (jtlayton) [inline qa/tasks/vstart_runner.py]: Shouldn't "pwd" be named "stdin"? That's what it ends up being, right?

File: qa/tasks/cephfs/mount.py [hunk 3]
```diff
@@ -111,6 +111,28 @@ def is_blacklisted(self):
                 return True
         return False
 
+    def create_file(self, filename='testfile', dirname=None, user=None,
+                    pwd=None, check_status=True):
```
Comment (jtlayton) [inline qa/tasks/cephfs/mount.py]: What purpose does "pwd" argument serve here?

File: qa/tasks/vstart_runner.py (hunk 2 above)
Comment (rishabh-d-dave) [reply qa/tasks/vstart_runner.py]: Done.

File: qa/tasks/cephfs/mount.py (hunk 3 above)
Comment (rishabh-d-dave) [reply qa/tasks/cephfs/mount.py]: No idea. Looks like I was supposed to remove it.

File: qa/tasks/vstart_runner.py (hunk 1 above)
Comment (rishabh-d-dave) [reply qa/tasks/vstart_runner.py]: Done.

**Comment by rishabh-d-dave** [comment]
@batrick @jtlayton 

Now that I am going again through all the commits after few weeks, my opinions of some commits have changed. I feel methods creating files should be dropped since it can be accomplished in single method call - `self.mount_a.testcmd('touch testfile')`.

**Comment by rishabh-d-dave** [comment]
There are 2 more improvements -

* I've covered the case where unlike bash, command like `sudo "ls ."` fail because Python's shell simulation doesn't recognize quotes as a token.
* Command to be executed as a different user should be a single string in the list.

File: qa/tasks/cephfs/mount.py (hunk 3 above)
Comment (batrick) [reply qa/tasks/cephfs/mount.py]: Are you going to? Let me know when your PR is done.

Comment (rishabh-d-dave) [reply qa/tasks/cephfs/mount.py]: I've removed it.

Can you please reply to this comment - https://github.com/ceph/ceph/pull/27481#issuecomment-482550708

**Comment by batrick** [comment]
> @batrick @jtlayton
> 
> Now that I am going again through all the commits after few weeks, my opinions of some commits have changed. I feel methods creating files should be dropped since it can be accomplished in single method call - `self.mount_a.testcmd('touch testfile')`.

If you're suggesting dropping all the create file methods in `mount.py`, I don't want us doing that -- too many unnecessary code changes.

**Comment by rishabh-d-dave** [comment]
What about dropping only the methods I've written? Commit bdd6bea5f4fd407372cc7e92bfba3db0ed1e163f and 052f62628a784e6edd1199006f17c9d4ab4f30b1 adds methods to create a file as a different user and with `check_status=False` respectively.

**Comment by batrick** [comment]
That's okay.

**Comment by rishabh-d-dave** [comment]
> That's okay.

Umm... sorry but I am a bit confused; does it mean "that's okay, drop them" or "that's okay, let those commits stay"?

Anyways, I spent some time testing this PR today and yesterday to test the new methods added in this PR. It looks OK to me now. If it's the case with you too, please proceed to merge it.  :)

**Comment by rishabh-d-dave** [comment]
While writing the last comment I realized that I could run test_acls.py too to test the new methods here. Although I must admit that I haven't looked at those tests since at least a month, I find them dependable because all tests were working fine until last time.

I tried that out and everything went fine - [test_acls.log](https://github.com/ceph/ceph/files/3091119/test_acls.log)

**Comment by batrick** [comment]
> Umm... sorry but I am a bit confused; does it mean "that's okay, drop them" or "that's okay, let those commits stay"?

either, I don't view it as too important. Merging this...

**Pull request merged**